simulation/simulation_mujoco.py

```diff
@@ -305,578 +197,338 @@
+    # and inertia matrix
+    mass_matrix = np.zeros((env.mjModel.nv, env.mjModel.nv))
+    mujoco.mj_fullM(env.mjModel, mass_matrix, env.mjData.qM)
+    # Get the mass matrix of the legs
+    legs_mass_matrix = LegsAttr(FL=mass_matrix[np.ix_(env.legs_qvel_idx.FL, env.legs_qvel_idx.FL)],
+                                FR=mass_matrix[np.ix_(env.legs_qvel_idx.FR, env.legs_qvel_idx.FR)],
+                                RL=mass_matrix[np.ix_(env.legs_qvel_idx.RL, env.legs_qvel_idx.RL)],
+                                RR=mass_matrix[np.ix_(env.legs_qvel_idx.RR, env.legs_qvel_idx.RR)])
+
+    for leg_id, leg_name in enumerate(legs_order):
+        if current_contact[leg_id] == 0:  # If in swing phase, compute the swing trajectory tracking control.
+            tau[leg_name], _, _ = stc.compute_swing_control(
+                model=env.mjModel,
+                q=qpos[env.legs_qpos_idx[leg_name]],
+                q_dot=qvel[env.legs_qvel_idx[leg_name]],
+                J=feet_jac[leg_name][:, env.legs_qvel_idx[leg_name]],
+                J_dot=jac_feet_dot[leg_name][:, env.legs_qvel_idx[leg_name]],
+                lift_off=lift_off_positions[leg_name],
+                touch_down=nmpc_footholds[leg_name],
+                swing_time=swing_time[leg_id],
+                foot_pos=feet_pos[leg_name],
+                foot_vel=feet_vel[leg_name],
+                h=legs_qfrc_bias[leg_name],
+                mass_matrix=legs_mass_matrix[leg_name]
+                )
+    # ---------------------------------------------------------------------------------------------------
+    # Set control and mujoco step ----------------------------------------------------------------------
+    # TODO: The order of the action space should not be hardoded, it should be provided by the environment.
+    action = np.concatenate((tau.to_list(order=["FR", "FL", "RR", "RL"]))).reshape(env.mjModel.nu)
+
+    env.step(action=action)
+
+    # Render only at a certain frequency
+    if time.time() - last_render_time > 1.0 / RENDER_FREQ:
+        feet_traj_geom_ids = plot_swing_mujoco(viewer=env.viewer,
+                                               swing_traj_controller=stc,
+                                               swing_period=swing_period,
+                                               swing_time=LegsAttr(FL=swing_time[0], FR=swing_time[1], RL=swing_time[2],
+                                                                   RR=swing_time[3]),
+                                               lift_off_positions=lift_off_positions,
+                                               nmpc_footholds=nmpc_footholds,
+                                               ref_feet_pos=ref_feet_pos,
+                                               geom_ids=feet_traj_geom_ids)
+        env.render()
+        last_render_time = time.time()
+
+    # Update the periodic gait generator
+    pgg.run(simulation_dt, pgg.step_freq)
+
+    if env.step_num > 2000:
```

Comment:
@giulioturrisi 

Here Giulio if I reset the environment the controller sometimes goes to hell. I think there are control specific variables that need to be resetted/reinitialized. Can you have a look please as you are more familiar with the controller?

Also it would be ideal if in the Controller class which will cover all posible controllers there is a .reset() method to deal with this in an Object Oriented fashion.